Add tests for xrefs on Python and JavaScript domains

agjohnson opened the pull request:
These are just the passing test cases for the domains currently. I am going to
patch up issues with nesting on both domains to start, so these are the test
cases I'll be testing against. I'll see about addressing the other core
domains, or at very least the cpp domain, with similar tests if this looks
okay.

So far, these tests only test against methods/functions for the basic nesting
structure. More complete tests will test additional domain roles.

Refs #662

agjohnson committed:
Add tests for xrefs on Python and JavaScript domains

agjohnson committed:
Reorder domain xref tests for legibility, clean up redundant tests

tk0miya commented:
Sorry, I can't understand what your testcase doing.
It looks like this confirms the argument of `JavaScriptDomain.find_obj()`.
I know this is useful to understand an internal of the domain. But it only confirms the code of the domain, not behavior. So it's not a test.

agjohnson commented:
Confirming the code of the domain is the point of these tests, as I intend to change the domains. This is confirming the arguments into `find_obj` finds and returns the intended object. The return from `find_obj` will contain the full name of the object we are looking up, or will be `None`, `None` on a miss.

agjohnson commented on tests/test_domain_py.py:
```diff
@@ -44,3 +45,52 @@ def test_function_signatures():
+            (u'NestedParentA.NestedChildA.subchild_2', (u'roles', u'method'))],
+        (u'NestedParentA.NestedChildA', u'NestedParentA.child_1', u'meth'): [
+            (u'NestedParentA.child_1', (u'roles', u'method'))],
+        (None, u'NestedChildA.subchild_1', u'meth'): [],
```
For instance, line 87 here is buggy behavior that I've patched with #3465. The arguments in to `find_obj` are not expected, and the return from `find_obj` is empty. Instead of:

```python
(None, u'NestedChildA.subchild_1', u'meth'): [],
```

This should instead be:

```python
(u'NestedParentA.NestedChildA', u'subchild_1', u'meth'): [
    (u'NestedParentA.NestedChildA.subchild_1', (u'roles', u'method'))
],
```

agjohnson commented on tests/test_domain_js.py:
```diff
@@ -0,0 +1,48 @@
+            # args = [domain, env, env object, role text, role type, order]
+            calls[args[2:-1]] = ret
+            return ret
+        return wrapped
```
Also, this could perhaps be simplified to call `app.env.domains['js'].find_obj` instead of mocking.

agjohnson committed:
Clean up testing implementation on Python and JavaScript domains

agjohnson committed:
Add mocked find_obj tests back in, reduce find_obj direct tests

The domains really need to be tested with a mocked find_obj and xrefs in the
source documents. Otherwise, the object classes could be tested directly, if
there was an easy pattern for testing directives.

tk0miya commented:
Thank you for reply and updating.
Okay, I understand the purpose and what the tests doing.

The following is summary of my understanding. Is this right?

* `test_domain_js_xrefs()` and `test_domain_py_xrefs()` are test cases to confirm how `find_obj()` is called from Sphinx app, builder, resolve and so on. In other words, this is test case for these modules, not for domains.
* `test_domain_js_objects()` and `test_domain_py_objects()` are test cases to confirm how the domain objects are stored.
* `test_domain_js_find_obj()` and `test_domain_py_find_obj()` are test cases to confirm how `find_obj()` works.

And I wonder the first one is really needed or not. I feel the subject of the test is ambiguous. In addition, it looks like fragile.
On the other hands, third one is surely important. +1.

Note: I don't have strong opinion for second one. I think it is already confirmed at third one. But it's okay if you needed it.

tk0miya commented on tests/test_domain_py.py:
```diff
@@ -44,3 +46,99 @@ def test_function_signatures():
+
+    assert find_obj(None, None, u'NONEXISTANT', u'class') == []
+    assert find_obj(None, None, u'NestedParentA', u'class') == [(
+        u'NestedParentA', (u'roles', u'class'))]
```
I feel this is a bit unreadable.
How about this?
```
    assert (find_obj(None, None, u'NestedParentA', u'class') ==
                     [(u'NestedParentA', (u'roles', u'class'))])
```
I like parenthesis are paired (preferably).
If so, we can add easily the case which `find_obj()` returns multiple objects.

agjohnson replied:
That looks better, I'll update.

agjohnson commented:
Sorry in advance for the long reply:

> test_domain_js_xrefs() and test_domain_py_xrefs() are test cases to confirm how find_obj() is called from Sphinx app, builder, resolve and so on. In other words, this is test case for these modules, not for domains.

These tests aim to verify that, given our test input document, object class prefixes and object names are correctly handled by `handle_signature`, `before_content`, and `after_content`. Specifically, these tests are aiming to verify that:

* `handle_signature` assumed the correct `py:class`/`py:module`/etc
* `handle_signature` returned the correct object and prefix
* `before_content` set the correct `py:class`/etc
* `after_content` handled reseting `py:class` correctly

We can ensure these cases because otherwise the calls to `find_obj` do not use the correct object name, or class prefix. 

> test_domain_js_objects() and test_domain_py_objects() are test cases to confirm how the domain objects are stored.

Not necessarily *how* they were stored, but *what* was stored. These tests ensure our object representation is what we were expecting, and are another test of `handle_signature`, `before_content`, and `after_content` handling of `py:class`/etc.

> test_domain_js_find_obj() and test_domain_py_find_obj() are test cases to confirm how find_obj() works.

This is correct, it's to ensure nothing unexpected happens when making changes in #3465 and #3475.

>And I wonder the first one is really needed or not. I feel the subject of the test is ambiguous. In addition, it looks like fragile.

The first tests are actually the most important to be testing when making changes to the domains. It tells us the most about the object nesting representation. It is testing a lot of pieces at once though, where perhaps the most correct approach here is to test the directive directly. It seems like this method could take a lot of mocking out of the application, environment, and object structure though, and I wasn't really sure where to start with this approach.

> On the other hands, third one is surely important. +1.
>
> Note: I don't have strong opinion for second one. I think it is already confirmed at third one. But it's okay if you needed it.

These two could be merged if need be. The second test was meant to test against *all* of the objects I expect to be created from the test doc. I don't necessarily need to test all of the output from `find_obj` for this however. I do want to test `find_obj` to ensure a few test cases work as expected for future changes however.

tk0miya commented:
>These tests aim to verify that, given our test input document, object class prefixes and object names are correctly handled by handle_signature, before_content, and after_content. 

That's strange. `find_obj()` is not related with these methods. It is only used to resolve references (`:py:*:` roles). I think `test_domain_js_objects()` is better to this purpose.

>It is testing a lot of pieces at once though, where perhaps the most correct approach here is to test the directive directly. 

Agreed. but there are no way to test directives directly. So we should test them through the results; like doctrees and domain datas (in this case, objects).

tk0miya commented on tests/test_domain_js.py:
```diff
@@ -0,0 +1,74 @@
+    assert objects['NestedParentB'] == ('roles', 'class')
+
+
+@pytest.mark.sphinx(testroot='domain-js')
```
By default, our test framework uses HTML builder by default. But, in these test cases, the HTML output is not needed. So it might be better to use dummy builder. It saves time for testing.
Could you check is it possible to replace?
```
@pytest.mark.sphinx('dummy', testroot='domain-js')
```

agjohnson commented:
> That's strange. find_obj() is not related with these methods. It is only used to resolve references (:py:*: roles). I think test_domain_js_objects() is better to this purpose.

`find_obj` doesn't use these methods directly, it uses the context put in place by them. `handle_signature` and `before_context` set the object class/etc prefixes, but I can't test these directive instances directly. The next place this context data is used from these objects is when linking is performed -- that is, when `find_obj` is ultimately called.

> Agreed. but there are no way to test directives directly. So we should test them through the results; like doctrees and domain datas (in this case, objects).

I'll give the doctree traversal pattern a try, especially if this is a more common pattern in tests. This is essentially the same thing as testing `find_obj` arguments though, just with more steps -- I'm ultimately testing the output of `find_obj` by looking at the doctree elements.

tk0miya commented:
Thank you for describing. I perfectly understood your idea (maybe).

>I'll give the doctree traversal pattern a try, especially if this is a more common pattern in tests. 

+1.
As a trial, I implemented it partially.
The test helper: `assert_pyrefnode()` is very similar to `assert_called()`. so it would be replaced easily. I hope this will help you.
```
diff --git a/tests/test_domain_py.py b/tests/test_domain_py.py
index ff8e2a4d..8187c9ba 100644
--- a/tests/test_domain_py.py
+++ b/tests/test_domain_py.py
@@ -16,6 +16,8 @@ import mock
 from sphinx import addnodes
 from sphinx.domains.python import py_sig_re, _pseudo_parse_arglist

+from util import assert_node
+

 def parse(sig):
     m = py_sig_re.match(sig)
@@ -55,16 +57,34 @@ def test_domain_py_xrefs(app, status, warning):
     app.env.domains['py'].find_obj = mock.Mock(wraps=find_obj)
     app.builder.build_all()

+    # roles.rst
+    doctree = app.env.get_doctree('roles')
+    refnodes = list(doctree.traverse(addnodes.pending_xref))
+
+    def assert_pyrefnode(node, domain, typ, module, cls, target):
+        attributes = {"refdomain": domain, "reftype": typ,
+                      "py:module": module, "py:class": cls, "reftarget": target}
+        assert_node(node, **attributes)
+
+    def assert_anyrefnode(node, target, **kwargs):
+        attributes = {"refdomain": "", "reftarget": target}
+        if 'module' in kwargs:
+            attributes['py:module'] = kwargs['module']
+        if 'cls' in kwargs:
+            attributes['py:class'] = kwargs['cls']
+        assert_node(node, **attributes)
+
+    assert_pyrefnode(refnodes[0], "py", "class", None, None, "TopLevel")
+    assert_pyrefnode(refnodes[1], "py", "meth", None, None, "top_level")
+    assert_pyrefnode(refnodes[2], "py", "meth", None, "NestedParentA", "child_1")
+    assert_pyrefnode(refnodes[3], "py", "meth", None, "NestedParentA", "NestedChildA.subchild_2")
+    assert_pyrefnode(refnodes[4], "py", "meth", None, "NestedParentA", "child_2")
+    assert_anyrefnode(refnodes[5], "any_child", cls="NestedParentA")
```

tk0miya commented:
Sorry, I had to mention you about branch.

Now the target branch of this PR is master. It's okay.
But it is better to merge #3465 to stable branch. Because it's bug fix.
So I propose you to move this to stable branch.

Note: on the other hand, #3475  is a new feature, so it should be merged into master branch.

agjohnson commented:
Closing this, I've opened up #3519 with feedback from here, and based on stable.
